src/ruby/pb/test/xds_client.rb

```diff
@@ -0,0 +1,193 @@
+$LOAD_PATH.unshift(pb_dir) unless $LOAD_PATH.include?(pb_dir)
+
+require 'optparse'
+require 'logger'
+
+require_relative '../../lib/grpc'
+require 'google/protobuf'
+
+require_relative '../src/proto/grpc/testing/empty_pb'
+require_relative '../src/proto/grpc/testing/messages_pb'
+require_relative '../src/proto/grpc/testing/test_services_pb'
+
+# Some global variables to be shared by server and client
+$watchers = Array.new
+$watchers_mutex = Mutex.new
+$current_request_id = 0;
+
+# RubyLogger defines a logger for gRPC based on the standard ruby logger.
+module RubyLogger
+  def logger
+    LOGGER
+  end
+
+  LOGGER = Logger.new(STDOUT)
+  LOGGER.level = Logger::INFO
+end
+
+# GRPC is the general RPC module
+module GRPC
+  # Inject the noop #logger if no module-level logger method has been injected.
+  extend RubyLogger
+end
+
+# creates a test stub
+def create_stub(opts)
+  address = "#{opts.server}"
+
+  GRPC.logger.info("... connecting insecurely to #{address}")
+  Grpc::Testing::TestService::Stub.new(
+    address,
+    :this_channel_is_insecure,
+  )
+end
+
+# This implements LoadBalancerStatsService required by the test runner
+class TestTarget < Grpc::Testing::LoadBalancerStatsService::Service
+  include Grpc::Testing
+
+  def get_client_stats(req, _call)
+    start_id = $current_request_id + 1
+    end_id = start_id + req['num_rpcs'] # watch the next X rpcs
+    start_time = Time.now.to_f
+    finish_time = start_time + req['timeout_sec']
+    index = 0
+    $watchers_mutex.synchronize do
+      index = $watchers.length
+      # Add a watcher for this range of id's
+      $watchers << {
+        "start_id" => start_id,
+        "end_id" => end_id,
+        "rpcs_by_peer" => Hash.new(0),
+        "rpcs_needed" => req['num_rpcs'],
+        "no_remote_peer" => 0
+      }
+    end
+    # wait until all RPCs are done, or we are past timeout
+    while ($watchers[index]['rpcs_needed'] > 0) &&
+          (Time.now.to_f < finish_time)
+      sleep(0.1)
+    end
+    LoadBalancerStatsResponse.new(
+      rpcs_by_peer: $watchers[index]['rpcs_by_peer'],
+      num_failures: $watchers[index]['no_remote_peer'] +
+      $watchers[index]['rpcs_needed']
+    );
+  end
+end
+
+# send 1 rpc every 1/qps second
+def run_test_loop(stub, duration_per_query)
+  include Grpc::Testing
+  req = SimpleRequest.new()
+
+  while true
+    start = Time.now.to_f
+    begin
+      resp = stub.unary_call(req)
+      remote_peer = resp.hostname
+    rescue GRPC::Unavailable => e
+      remote_peer = ""
+    end
+    finish = Time.now.to_f
+    delay = duration_per_query - (finish - start)
+    if delay > 0
+      sleep(delay)
+    else
+      GRPC.logger.info("ruby xds: warning, rpc takes too long to finish. " \
+                       "If you consistently see this, the qps is too high.")
+    end
+    $current_request_id += 1
+    $watchers_mutex.synchronize do
+      $watchers.each do |watcher|
+        if (watcher['start_id'] <= $current_request_id) &&
```

Comment:
nit: if we remove `watchers` from the `$watchers` list after we're done with them, then I think there's no need to have the `end_id` field within watcher objects at all.

I think we can delete this conditional logic then and unconditionally run the body of this if statement

Reply:
Yes this works. Basically the idea is that when we add a `watcher`, we are interested in the _next_ `num_rpcs`. So each subsequent completed rpcs will just decrement every `watcher`'s `num_rpcs` in the list by 1. When a `watcher` reaches `num_rpcs = 0` (or timed out), it's removed from the `watchers` list. So it can be reasoned that any successful rpc is useful to _every_ `watcher` currently still in the list.